sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -932,6 +933,33 @@ object PushPredicateThroughGenerate extends Rule[LogicalPlan] with PredicateHelp
 }
 
 /**
+ * Push [[Filter]] operators through [[Window]] operators. Parts of the predicate that can be pushed
+ * beneath must satisfy three conditions:
+ * 1. involving one and only one column that is part of window partitioning key.
+ * 2. Window partitioning key should be just a sequence of [[AttributeReference]].
```

Comment:
I think I might know the reason. When the partitioning key is like `k1 + k2`, if we push down the predicates `k1 + 2`, it will change the rows in the window.